src/OpenTelemetry/Trace/Export/BatchingSpanProcessor.cs

```diff
@@ -0,0 +1,215 @@
+                    Debug.WriteLine($"!!! {DateTime.UtcNow:o}  export " + this.currentQueueSize);
+                    await this.ExportBatchAsync(cancellationToken).ConfigureAwait(false);
+                    Debug.WriteLine($"!!! {DateTime.UtcNow:o}  export finished" + this.currentQueueSize);
+                }
+
+                // there is no point in waiting for a worker task if cancellation happens
+                // it's dead already or will die on the next iteration on its own
+
+                // ExportBatchAsync must never throw, we are here either because it was cancelled
+                // or because there are no items left
+                OpenTelemetrySdkEventSource.Log.ShutdownEvent(this.currentQueueSize);
+            }
+        }
+
+        public void Dispose()
+        {
+            if (!this.stopping)
+            {
+                this.ShutdownAsync(CancellationToken.None).ContinueWith(_ => { }).GetAwaiter().GetResult();
+            }
+        }
+
+        private async Task ExportBatchAsync(CancellationToken cancellationToken)
+        {
+            try
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return;
+                }
+
+                var nextBatchSize = Math.Min(this.currentQueueSize, this.maxExportBatchSize);
+
+                if (nextBatchSize == 0)
+                {
+                    return;
+                }
+
+                var batch = new List<Span>(nextBatchSize);
+
+                while (batch.Count < nextBatchSize && this.exportQueue.TryDequeue(out var nextSpan))
+                {
+                    Interlocked.Decrement(ref this.currentQueueSize);
+                    batch.Add(nextSpan);
+                }
+
+                var result = await this.Exporter.ExportAsync(batch, cancellationToken).ConfigureAwait(false);
+                if (result != SpanExporter.ExportResult.Success)
+                {
+                    OpenTelemetrySdkEventSource.Log.ExporterErrorResult(result);
+
+                    // we do not support retries for now and leave it up to exporter
+                    // as only exporter implementation knows how to retry: which items failed
+                    // and what is the reasonable policy for that exporter.
+                }
+            }
+            catch (Exception ex)
+            {
+                OpenTelemetrySdkEventSource.Log.SpanProcessorException(ex);
+            }
+        }
+
+        private async Task Worker(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                var sw = Stopwatch.StartNew();
+                await this.ExportBatchAsync(cancellationToken).ConfigureAwait(false);
+
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return;
+                }
+
+                var remainingWait = this.scheduleDelay - sw.Elapsed;
```

Comment:
Do we want a delay to happen even if the queue has items and the exporter is able to flush more spans?
Would some sort of back-pressure control be implemented instead? Perhaps the exporter could `await Task.Delay` instead, if it needs to called less often (i.e it gets a HTTP 429 from the backend, `Retry-After` would tell it how long to wait.

Reply:
processor must not deal with the http codes (it doesn't even know whether it was HTTP). exporter could only tell if error eas transient.
So my understanding is that exporter has to do all reasonable retries and only exporter knows which retry policy this backend needs. 

Leaving retries topic: spec does not tell anything about reducing delay depending on the queue size. While it makes total sense to me, I'd defer the complications in the processor to the moment we get real near-production users feedback. My worry is that this behavior would not be generic for all exporters.

Now users could only try to tune delay manually for their case.

Reply:
Right. The HTTP example I mentioned would be in the exporter: `exporter could await Task.Delay if it gets back-off from it's backend, i.e 429, in a hypothetical HTTP exporter)`
Happy to not deal with time delay right now. It indeed makes sense to define this at a spec level.